main/src/ca/uwaterloo/flix/api/lsp/provider/Visitors.scala

```diff
@@ -0,0 +1,316 @@
+    // visit(Entity.Root(root))?
+
+    root.defs.map{ case (_, defn) => {
+      if (inside(accept, defn)) {
+        visitDef(visit, accept)(defn)
+      }
+    }}
+
+    root.effects.map{ case (_, v) => {
+      if (inside(accept, v)) {
+        visitEffect(visit, accept)(v)
+      }
+    }}
+
+    // root.entryPoint.map{ case v => visitEntryPoint(visit, accept)(v) }
+    // root.enums.map{ case (_, v) => visitEnum(???, ???)(v) };
+    // root.instances.map { case (_, v) => visitInstance(???, ???)(v) };
+    // root.modules.map { case (_, v) => visitModule(???, ???)(v) };
+    // root.restrictableEnums.map { case (_, v) => visitRestrictable(???, ???)(v) }; //experimental, prob should be removed
+    // root.sigs.map{ case (_, v) => visitSig(???, ???)(v) };
+    // root.structs.map{ case (_, v) => visitStruct(???, ???)(v) };
+    // root.traitEnv.map{ case (_, v) => visitTraitEnv(???, ???)(v) };
+    // root.traits.map{ case (_, v) =>  visitTrait(???, ???)(v) };
+    // root.typeAliases.map{ case (_, v) => visitTypeAlias(???, ???)(v) };
+    // root.uses.map{ case (_, v) => visitUse(???, ???)(v) };
+  }
+
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
+    val posLine = pos.line + 1
+    val posCol = pos.character + 1
+
+    // sp1 and sp2, by invariant, has the same source, so we can use either
+    (uri == loc.sp1.source.name) &&
+    (posLine >= loc.beginLine) &&
+    (posLine <= loc.endLine) &&
+    !(posLine == loc.beginLine && posCol < loc.beginCol) &&
+    !(posLine == loc.endLine && posCol > loc.endCol)
+  }
+
+  def inside(other: SourceLocation => Boolean, defn: Def): Boolean = {
+    other(defn.spec.loc) || other(defn.exp.loc) || other(defn.sym.loc)
+  }
+
+  def inside(other: SourceLocation => Boolean, expr: Expr): Boolean = other(expr.loc)
+
+  def inside(other: SourceLocation => Boolean, effect: Effect): Boolean = {
+    other(effect.loc) || effect.ann.annotations.exists(ann => inside(other, ann)) || other(effect.sym.loc)
+  }
+
+  def inside(other: SourceLocation => Boolean, ann: Annotation): Boolean = other(ann.loc)
+
+  def inside(loc1: SourceLocation, loc2: SourceLocation): Boolean = {
+    loc1.source == loc2.source &&
+    (loc1.beginLine >= loc2.beginLine) &&
+    (loc1.endLine <= loc2.endLine) &&
+    !(loc2.beginLine == loc1.beginLine && loc2.beginCol > loc1.beginCol) &&
+    !(loc1.endLine == loc2.endLine && loc1.endCol > loc2.endCol)
+  }
+
+  def visitEffect(visit: Entity => Unit, accept: SourceLocation => Boolean)(effect: Effect): Unit = {
+    visit(Entity.Effect(effect))
+    ???
+  }
+
+  def visitDef(visit: Entity => Unit, accept: SourceLocation => Boolean)(defn: Def): Unit = {
+    visit(Entity.Def(defn))
+    if (accept(defn.spec.loc)) {
+      ???
+    }
+
+    if (inside(accept, defn.exp)) {
+      visitExpr(visit, accept)(defn.exp)
+    }
+  }
+
+  def visitExpr(visit: Entity => Unit, accept: SourceLocation => Boolean)(expr: Expr): Unit = {
```

Comment:
I know its tempting to use Entity, but I think that abstraction is broken too.

Just go straight for the ball and use Expr.

Reply:
This means that you will need multiple visit functions at the top, e.g. both acceptExpr, acceptDefn, acceptBlah. But dont worry, later we can put them behind an interface. For now just make the signatures correct (even if long).

Reply:
got it